src/renderer/base/renderer.cpp

```diff
@@ -664,6 +664,71 @@ void Renderer::_PaintBufferOutputHelper(_In_ IRenderEngine* const pEngine,
     }
 }
 
+void Renderer::_PaintBufferLineOutputHelper(_In_ IRenderEngine* const pEngine,
+                                            const ROW& row,
+                                            const COORD target)
+{
+    size_t cols = 0;
+    const ATTR_ROW& attrRow = row.GetAttrRow();
+    const std::vector<Cluster> rowClusters = row.GetClusters();
+
+    const size_t runCount = attrRow.GetNumberOfRuns();
+    size_t runIndex = 0;
+    // And hold the point where we should start drawing.
+    auto screenPoint = target;
+
+    size_t runStart = 0;
+    // This outer loop will continue until we reach the end of the text we are trying to draw.
+    while (runIndex < runCount)

```

Comment:
This is where magic happens. We have already processed the text buffer and got the clusters we need. All that left is to walk through the row and render each run.